:sparkles: [e2e framework] Add ability to run pre and post actions during clusterctl upgrade spec

detiber opened the pull request:
**What this PR does / why we need it**:

Allows a provider to specify pre and post upgrade actions to take when running ClusterctlUpgradeSpec

Also enable better cleanup in clusterctl upgrade spec when management
cluster hasn't been upgraded yet.

detiber commented on test/e2e/clusterctl_upgrade.go:
```diff
@@ -263,6 +277,22 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
+					}, input.E2EConfig.GetIntervals(specName, "wait-delete-cluster")...)
+				} else {
+					log.Logf("Management Cluster does not appear to support CAPI resources.")
+				}
```
This cleanup step would previously fail when the spec fails prior to upgrading the management cluster, this allows cleanup to proceed if failure occurs prior to running `clusterctl upgrade`, or even before `clusterctl init` has been run.

detiber commented on test/e2e/clusterctl_upgrade.go:
```diff
@@ -52,6 +54,8 @@ type ClusterctlUpgradeSpecInput struct {
 	BootstrapClusterProxy framework.ClusterProxy
 	ArtifactFolder        string
 	SkipCleanup           bool
+	PreUpgrade            func(managementClusterProxy framework.ClusterProxy)
+	PostUpgrade           func(managementClusterProxy framework.ClusterProxy)
```
This allows for pre and post upgrade steps to be run by providers when running this spec.

For example, in cluster-api-provider-packet, we require users to execute both pre and post upgrade actions when moving from v1alpha3 to v1alpha4. This is to enable the use of a newer version of our cloud provider implementation, which needs to be done prior to the v1alpha4 upgrade, and an additional step to update the providerIDs on existing nodes after the v1alpha4 upgrade is finished.

vincepri replied:
Thanks for this @detiber! (also cc @CecileRobertMichon which might be interested in this from CAPZ perspective)

> and an additional step to update the providerIDs on existing nodes after the v1alpha4 upgrade is finished

Curious about this one, do you have more details?

CecileRobertMichon replied:
@vincepri I don't think this applies to CAPZ,  but no objections from my side to enabling it for other providers as long as it's optional and doesn't break existing tests

sbueringer commented on test/e2e/clusterctl_upgrade.go:
```diff
@@ -219,6 +223,11 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 		By("THE MANAGEMENT CLUSTER WITH OLDER VERSION OF PROVIDERS WORKS!")
 
+		if input.PreUpgrade != nil {
+			By("Running Pre-upgrade steps against the amnagement cluster")
```
```suggestion
			By("Running Pre-upgrade steps against the management cluster")
```
nit, typo (same below)

sbueringer commented on test/e2e/clusterctl_upgrade.go:
```diff
@@ -263,6 +277,22 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 
 			if !input.SkipCleanup {
+				if discovery.ServerSupportsVersion(managementClusterProxy.GetClientSet().DiscoveryClient, clusterv1.GroupVersion) == nil {
+					Byf("Deleting all clusters in Namespace: %s in management cluster: %s", testNamespace.Name, managementClusterName)
```
```suggestion
					Byf("Deleting all clusters in namespace: %s in management cluster: %s", testNamespace.Name, managementClusterName)
```
nit (same below)

\+ golangci-lint finding

sbueringer commented on test/e2e/clusterctl_upgrade.go:
```diff
@@ -299,3 +329,73 @@ func downloadToTmpFile(url string) string {
 	return tmpFile.Name()
 }
+
+// deleteAllClustersAndWait deletes all v1alpha3 cluster resources in the given namespaceand waits for them to be gone.
```
```suggestion
// deleteAllClustersAndWait deletes all v1alpha3 cluster resources in the given namespace and waits for them to be gone.
```
nit

sbueringer reviewed (commented):
Looks great, thx :) A few nits from my side

detiber commented:
@sbueringer thanks for the review, I believe I've addressed all the comments/suggestions as well as the golangci-lint issue.

As an aside, I actually thought about pushing the old/current api support into the framework properly, but it would have required much more invasive changes that would have had bigger impact to the API (either breaking changes to existing method signatures or a more in depth migration and deprecation strategy). I suspect it could be something valuable that we might want to think about, but would warrant a much bigger discussion :)

fabriziopandini commented on test/e2e/clusterctl_upgrade.go:
```diff
@@ -299,3 +330,73 @@ func downloadToTmpFile(url string) string {
 }
+
+// deleteAllClustersAndWait deletes all v1alpha3 cluster resources in the given namespace and waits for them to be gone.
+func deleteAllClustersAndWait(ctx context.Context, input framework.DeleteAllClustersAndWaitInput, intervals ...interface{}) {
```
Q: should this be implemented as a helper in https://github.com/kubernetes-sigs/cluster-api/blob/master/test/framework/cluster_helpers.go (so we can benefit from `GetAllClustersByNamespace`, `DeleteCluster`, `WaitForClusterDeleted` already existing there)?

sbueringer replied:
Can we use those? The new funcs are all using v1alpha3 types which we kind of only need here.

detiber replied:
@fabriziopandini more than happy to move it to the framework, my initial thought was that we could add the `discovery.ServerSupportsVersion()` check into the main helper functions, that way we could even use most of those helpers in the pre-upgrade steps in the clusterctl_upgrade test, but the existing method signatures complicated that a bit.

I'm more than happy to try and explore either that route or just add parallel methods where needed for the "old" api version. Just let me know which approach you prefer.

fabriziopandini replied:
This is a good point.
What about adding a v1alpa3 suffix to those func so it is more evident why we are using custom helpers here

sbueringer replied:
@detiber @fabriziopandini 

Just another idea. What about implementing version-independent deletion helpers? Looks like we only access a few fields during deletion, so I guess it can be reasonably easy implemented with unstructured.

(Feel free to ignore if that would be too hacky :)

detiber replied:
@sbueringer I'm going to experiment with doing version independence with the framework helpers, but will separate that work out into it's own PR if it works out.

sbueringer replied:
Sounds good to me. I didn't see your last answer before sending mine :)

fabriziopandini reviewed (commented):
This is a nice addition!
lgtm pending a question on the additional Cluster helpers

sbueringer commented on test/e2e/clusterctl_upgrade.go:
```diff
@@ -299,3 +330,73 @@ func downloadToTmpFile(url string) string {
+	return clusters
+}
+
+// deleteClusterOldAPIInput is the input for deleteCluster.
```
```suggestion
// deleteClusterOldAPIInput is the input for deleteClusterOldAPIInput.
```

sbueringer commented on test/e2e/clusterctl_upgrade.go:
```diff
@@ -299,3 +330,73 @@ func downloadToTmpFile(url string) string {
+
+// deleteAllClustersAndWaitOldAPI deletes all v1alpha3 cluster resources in the given namespace and waits for them to be gone.
+func deleteAllClustersAndWaitOldAPI(ctx context.Context, input framework.DeleteAllClustersAndWaitInput, intervals ...interface{}) {
+	Expect(ctx).NotTo(BeNil(), "ctx is required for DeleteAllClustersAndWait")
+	Expect(input.Client).ToNot(BeNil(), "Invalid argument. input.Client can't be nil when calling DeleteAllClustersAndWait")
+	Expect(input.Namespace).ToNot(BeEmpty(), "Invalid argument. input.Namespace can't be empty when calling DeleteAllClustersAndWait")
```
```suggestion
	Expect(ctx).NotTo(BeNil(), "ctx is required for deleteAllClustersAndWaitOldAPI")
	Expect(input.Client).ToNot(BeNil(), "Invalid argument. input.Client can't be nil when calling deleteAllClustersAndWaitOldAPI")
	Expect(input.Namespace).ToNot(BeEmpty(), "Invalid argument. input.Namespace can't be empty when calling deleteAllClustersAndWaitOldAPI")
```

sbueringer commented:
/lgtm

fabriziopandini reviewed (commented):
/lgtm
/approve
